plugins/modules/vios_update_upgrade.py

```diff
@@ -0,0 +1,502 @@
+    m_system = attributes['system_name']
+    sys_list = (
+        hmc_conn.execute("lssyscfg -r sys -F name").splitlines() + hmc_conn.execute("lssyscfg -r sys -F type_model*serial_num").splitlines()
+    )
+    if m_system not in sys_list:
+        module.fail_json(msg="The managed system is not available in HMC")
+    else:
+        if attributes['vios_name'] is not None:
+            vios_list = list(hmc_conn.execute("lssyscfg -r lpar -m {0} -F name".format(m_system)).splitlines())
+        elif attributes['vios_id'] is not None:
+            vios_list = list(hmc_conn.execute("lssyscfg -r lpar -m {0} -F lpar_id".format(m_system)).splitlines())
+        if vios_name not in vios_list:
+            module.fail_json(msg="The vios is not available in the managed system")
+    if attributes['vios_name'] is not None:
+        version = hmc_conn.execute("viosvrcmd -p {0} -m {1} -c ioslevel".format(vios_name, m_system)).strip()
+    elif attributes['vios_id'] is not None:
+        version = hmc_conn.execute("viosvrcmd --id {0} -m {1} -c ioslevel".format(vios_name, m_system)).strip()
+    version = {
+        "vios": vios_name,
+        "system": m_system,
+        "version": version}
+    return False, version, None
+
+
+def ensure_update_upgrade(module, params):
+    hmc_host = params['hmc_host']
+    hmc_user = params['hmc_auth']['username']
+    password = params['hmc_auth']['password']
+    validate_parameters(params)
+    attributes = params.get('attributes')
+    hmc_conn = HmcCliConnection(module, hmc_host, hmc_user, password)
+    hmc = Hmc(hmc_conn)
+
+    vios_name = attributes['vios_name'] or attributes['vios_id']
+    m_system = attributes['system_name']
+    sys_list = (
+        hmc_conn.execute("lssyscfg -r sys -F name").splitlines() + hmc_conn.execute("lssyscfg -r sys -F type_model*serial_num").splitlines()
+    )
+    if m_system not in sys_list:
+        module.fail_json(msg="The managed system is not available in HMC")
+    else:
+        if attributes['vios_name'] is not None:
+            vios_list = list(hmc_conn.execute("lssyscfg -r lpar -m {0} -F name".format(m_system)).splitlines())
+        elif attributes['vios_id'] is not None:
+            vios_list = list(hmc_conn.execute("lssyscfg -r lpar -m {0} -F lpar_id".format(m_system)).splitlines())
+        if vios_name not in vios_list:
+            module.fail_json(msg="The vios is not available in the managed system")
+
+    if attributes['repository'] in ['nfs', 'sftp']:
+        if attributes['save'] is not None and attributes['image_name'] is None:
+            raise ParameterError("To save the image to the HMC hard disk, 'image_name' parameter is required")
+        if attributes['save'] is None and attributes['image_name'] is not None:
+            raise ParameterError("For remote server repository'image_name' parameter is only required if 'save' option is set to 'true'")
+
+    files = ''
+    if attributes['files'] is not None:
```

Comment:
`",".join(attributes['files'])` - does this suffice?

Reply:
Done